File: run_demo.py
#!/usr/bin/env python3
"""
CLI demo -- runs a few simulated days through the whole pipeline and prints
a summary. Zero external dependencies (stdlib only). Good smoke test after
touching core/, and a quick way to see the system work without the UI.

Usage:
    python3 run_demo.py [num_days]
"""
from __future__ import annotations

import sys

from core import library


def main() -> None:
    days = int(sys.argv[1]) if len(sys.argv) > 1 else 3
    library.reset_all()

    print(f"IECHM Lead Generation Engine -- running {days} simulated day(s)\n")

    for _ in range(days):
        summary = library.run_cycle()
        print(f"=== Day {summary['day']} ===")
        print(f"  raw signals ingested : {summary['total_signals']}")
        print(f"  qualified leads      : {summary['total_qualified']}")
        print(f"  bids approved        : {summary['total_bids_approved']}")
        print(f"  sentinel events      : {summary['sentinel_events_this_cycle']}")
        for ch in summary["channels"]:
            print(
                f"    - {ch['channel_type']:<24} signals={ch['signals']:<4} "
                f"qualified={ch['qualified']:<4} bids_ok={ch['bids_approved']:<3} "
                f"health={ch['channel_health_score']}"
            )
            if ch["subdomain_agents_spawned"]:
                print(f"        spawned L4 workers: {ch['subdomain_agents_spawned']}")
            if ch["subdomain_agents_deprecated"]:
                print(f"        deprecated L4 workers: {ch['subdomain_agents_deprecated']}")
        print()

    print("=== Funnel (cumulative) ===")
    funnel = library.funnel_metrics()
    for k, v in funnel["totals"].items():
        print(f"  {k:<16}: {v}")

    print("\n=== Agent swarm ===")
    agents = library.list_agents()
    by_layer: dict[str, int] = {}
    for a in agents:
        by_layer[a["layer"]] = by_layer.get(a["layer"], 0) + 1
    for layer, count in sorted(by_layer.items()):
        print(f"  {layer:<24}: {count}")
    print(f"  TOTAL AGENTS: {len(agents)}")

    events = library.list_sentinel_events(limit=10)
    print(f"\n=== Sentinel events (most recent {len(events)}) ===")
    for e in events:
        print(f"  [{e['event_type']}] agent={e['agent_id']} drift={e['drift_score']:.2f} :: {e['reason']}")

    print("\n=== Strategy ledger ===")
    for s in library.list_strategies():
        print(f"  {s['strategy_id']:<45} wins={s['wins']:<3} losses={s['losses']:<3} status={s['status']}")

    print("\n=== Sample qualified lead (full pipeline trace) ===")
    leads = library.list_leads(limit=1, status="BID_APPROVED")
    if leads:
        detail = library.get_lead(leads[0]["lead_id"])
        lead = detail["lead"]
        bid = detail["bid"]
        print(f"  Platform     : {lead['platform']} / {lead['sub_domain'] or '-'}")
        print(f"  Title        : {lead['title']}")
        print(f"  Archetype    : {lead['client_archetype']}")
        print(f"  Domain       : {lead['domain']}")
        print(f"  Market price : ${lead['market_price_usd']:,.2f}")
        print(f"  Bid price    : ${lead['bid_price_usd']:,.2f}  (margin {lead['margin_pct']:.1f}%)")
        if bid:
            print(f"  Strategy     : {bid['strategy_name']} (explore={bool(bid['explore'])})")
            print(f"  Reviewer     : {bid['reviewer_status']}")
            print(f"  --- proposal ---\n{bid['proposal_text']}\n")

    print(f"\nDB written to: core/db.py -> {library.db.DB_PATH}")
    print("Start the dashboard with:  python3 server.py")


if __name__ == "__main__":
    main()

File: iechm-leadgen-engine-c/src/run_demo.c
/* CLI smoke test -- mirrors run_demo.py. Talks to store.c/orchestrator.c
 * directly rather than through library.c's JSON strings: library.c exists
 * to serve http.c, and a CLI printing typed fields has no reason to
 * serialize to JSON and back. Persists to the same iechm.db next to the
 * binary that the server uses, so a demo run and the dashboard can look at
 * the same data, exactly like the original (both go through core.library).
 *
 * Usage: ./iechm_demo [num_days]
 */
#include "store.h"
#include "orchestrator.h"
#include "models.h"
#include "fmt.h"
#include "rng.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static void get_exe_dir(char *out, size_t cap) {
    char path[4096];
    ssize_t n = readlink("/proc/self/exe", path, sizeof path - 1);
    if (n <= 0) { xcpy(out, cap, "."); return; }
    path[n] = '\0';
    char *slash = strrchr(path, '/');
    if (slash) *slash = '\0';
    xcpy(out, cap, path);
}

int main(int argc, char **argv) {
    global_rng_seed();
    int days = argc > 1 ? atoi(argv[1]) : 3;

    char exe_dir[2048], db_path[2048];
    get_exe_dir(exe_dir, sizeof exe_dir);
    snprintf(db_path, sizeof db_path, "%s/iechm.db", exe_dir);

    Store store;
    store_init(&store);
    Orchestrator orch;
    orchestrator_init(&orch);
    store_save_to_disk(&store, db_path); /* reset_all()'s effect: start from empty */

    printf("IECHM Lead Generation Engine -- running %d simulated day(s)\n\n", days);

    for (int d = 1; d <= days; d++) {
        CycleSummary summary;
        orchestrator_run_cycle(&orch, &store, d, &summary);
        store_save_to_disk(&store, db_path);

        printf("=== Day %d ===\n", summary.day);
        printf("  raw signals ingested : %d\n", summary.total_signals);
        printf("  qualified leads      : %d\n", summary.total_qualified);
        printf("  bids approved        : %d\n", summary.total_bids_approved);
        printf("  sentinel events      : %d\n", summary.sentinel_events_this_cycle);
        for (int i = 0; i < summary.channels_count; i++) {
            const ChannelCycleSummary *ch = &summary.channels[i];
            printf("    - %-24s signals=%-4d qualified=%-4d bids_ok=%-3d health=%g\n",
                   ch->channel_type, ch->signals, ch->qualified, ch->bids_approved, ch->channel_health_score);
            if (ch->spawned_count) {
                printf("        spawned L4 workers: [");
                for (int j = 0; j < ch->spawned_count; j++) printf("%s'%s'", j ? ", " : "", ch->spawned[j]);
                printf("]\n");
            }
            if (ch->deprecated_count) {
                printf("        deprecated L4 workers: [");
                for (int j = 0; j < ch->deprecated_count; j++) printf("%s'%s'", j ? ", " : "", ch->deprecated[j]);
                printf("]\n");
            }
        }
        printf("\n");
    }

    printf("=== Funnel (cumulative) ===\n");
    long long raw_signals = 0, layer0 = 0, qualified = 0, approved = 0, rejected = 0;
    for (int i = 0; i < store.run_log.count; i++) {
        const RunLogEntry *e = vec_at(&store.run_log, i);
        raw_signals += e->signals_count;
        layer0 += e->layer0_passed;
        qualified += e->qualified;
        approved += e->bids_approved;
        rejected += e->bids_rejected;
    }
    printf("  %-16s: %lld\n", "raw_signals", raw_signals);
    printf("  %-16s: %lld\n", "layer0_passed", layer0);
    printf("  %-16s: %lld\n", "qualified", qualified);
    printf("  %-16s: %lld\n", "bids_approved", approved);
    printf("  %-16s: %lld\n", "bids_rejected", rejected);

    printf("\n=== Agent swarm ===\n");
    for (AgentLayer l = AL_L1_CENTRAL_COMMAND; l <= AL_SENTINEL; l++) {
        int count = 0;
        for (int i = 0; i < store.agents.count; i++) {
            const StoredAgent *a = vec_at(&store.agents, i);
            if (strcmp(a->layer, agent_layer_str(l)) == 0) count++;
        }
        if (count) printf("  %-24s: %d\n", agent_layer_str(l), count);
    }
    printf("  TOTAL AGENTS: %d\n", store.agents.count);

    int to_show = store.sentinel_events.count < 10 ? store.sentinel_events.count : 10;
    printf("\n=== Sentinel events (most recent %d) ===\n", to_show);
    for (int i = store.sentinel_events.count - 1, shown = 0; i >= 0 && shown < 10; i--, shown++) {
        const StoredSentinelEvent *e = vec_at(&store.sentinel_events, i);
        printf("  [%s] agent=%s drift=%.2f :: %s\n", e->event_type, e->agent_id, e->drift_score, e->reason);
    }

    printf("\n=== Strategy ledger ===\n");
    int ns = store.strategies.count;
    int *idx = ns ? malloc(sizeof(int) * (size_t) ns) : NULL;
    for (int i = 0; i < ns; i++) idx[i] = i;
    for (int i = 1; i < ns; i++) {
        int key = idx[i];
        const StoredStrategy *ka = vec_at(&store.strategies, key);
        int j = i - 1;
        while (j >= 0) {
            const StoredStrategy *ja = vec_at(&store.strategies, idx[j]);
            int c = strcmp(ja->domain, ka->domain);
            if (c == 0) c = strcmp(ja->name, ka->name);
            if (c <= 0) break;
            idx[j + 1] = idx[j];
            j--;
        }
        idx[j + 1] = key;
    }
    for (int i = 0; i < ns; i++) {
        const StoredStrategy *s = vec_at(&store.strategies, idx[i]);
        printf("  %-45s wins=%-3d losses=%-3d status=%s\n", s->strategy_id, s->wins, s->losses, s->status);
    }
    free(idx);

    printf("\n=== Sample qualified lead (full pipeline trace) ===\n");
    const StoredLead *sample = NULL;
    for (int i = store.leads.count - 1; i >= 0; i--) {
        const StoredLead *l = vec_at(&store.leads, i);
        if (strcmp(l->status, "BID_APPROVED") == 0) { sample = l; break; }
    }
    if (sample) {
        const StoredSignal *sig = NULL;
        for (int i = 0; i < store.signals.count; i++) {
            const StoredSignal *s = vec_at(&store.signals, i);
            if (strcmp(s->signal_id, sample->signal_id) == 0) { sig = s; break; }
        }
        const StoredBid *bid = NULL;
        for (int i = store.bids.count - 1; i >= 0; i--) {
            const StoredBid *b = vec_at(&store.bids, i);
            if (strcmp(b->lead_id, sample->lead_id) == 0) { bid = b; break; }
        }

        printf("  Platform     : %s / %s\n", sig ? sig->platform : "", (sig && sig->sub_domain[0]) ? sig->sub_domain : "-");
        printf("  Title        : %s\n", sig ? sig->title : "");
        printf("  Archetype    : %s\n", sample->client_archetype);
        printf("  Domain       : %s\n", sample->domain);
        char market_str[48], bid_str[48];
        format_money_commas(sample->market_price_usd, market_str, sizeof market_str);
        format_money_commas(sample->bid_price_usd, bid_str, sizeof bid_str);
        printf("  Market price : $%s\n", market_str);
        printf("  Bid price    : $%s  (margin %.1f%%)\n", bid_str, sample->margin_pct);
        if (bid) {
            printf("  Strategy     : %s (explore=%s)\n", bid->strategy_name, bid->explore ? "True" : "False");
            printf("  Reviewer     : %s\n", bid->reviewer_status);
            printf("  --- proposal ---\n%s\n\n", bid->proposal_text);
        }
    }

    printf("\nDB written to: %s\n", db_path);
    printf("Start the dashboard with:  ./iechm_server\n");
    return 0;
}
